ast/format.go

```diff
@@ -154,18 +179,37 @@ func (f *formatter) formatChildWithParens(parent, child Node) {
 		pvc = getPrecedence(index)
 	}
 
-	// If one of parent or child has not matched any case, then we shouldn't apply any parenthesis.
+	return pvp, pvc
+}
+
+// About parenthesis:
+// We need parenthesis if a child node has lower precedence (bigger value) than its parent node.
+// The same stands for the left child of a binary expression; while, for the right child, we need parenthesis if its
+// precedence is lower or equal then its parent's.
+//
+// To explain parenthesis logic, we must to understand how the parser generates the AST.
+// (A) - The parser always puts lower precedence operators at the root of the AST.
+// (B) - When there are multiple operators with the same precedence, the right-most expression is at root.
+// (C) - When there are parenthesis, instead, the parser recursively generates a AST for the expression contained
+// in the parenthesis, and makes it the right child.
+// So, when formatting:
+//  - if we encounter a child with lower precedence on the left, this means it requires parenthesis, because, for sure,
+//    the parser detected parenthesis to break (A);
+//  - if we encounter a child with higher or equal precedence on the left, it doesn't need parenthesis, because
+//    that was the natural parsing order of elements (see (B));
+//  - if we encounter a child with lower or equal precedence on the right, it requires parenthesis, otherwise, it
+//    would have been at root (see (C)).
+func needsPar(pvp, pvc int, isRight bool) bool {
```

Comment:
Can you rename this something that's more descriptive? Maybe just spell out the entire word.